### ISSUE TITLE
Mapping over empty list

### ISSUE THREAD

**Issue by lukejagodzinski** [samdenty/gqless#63]
There is a problem with mapping an empty array of values. Let's take the hackernews example. If I query the following data:

```ts
<div>
  {query.hn!.topStories({ limit: 0 })!.map(story =>
    <Story key={story!.id} story={story!} />
  )}
</div>
```

It will still try to render one element because the `story` element is a proxy object.
Here is the reproduction repo: https://codesandbox.io/s/gqless-empty-list-rprc5

I thought that maybe using the `ofType` function would help but not. I guess it would be good to add some helper function that checks if a value in the proxy is loaded? We shouldn't even try to render an element if a value is not there. Or maybe proxies for arrays should work differently?

EDIT:
What I think is wrong, is that `query.hn!.topStories({ limit: 0 })!.length` always returns `1` for empty list

EDIT2:
I guess in this line https://github.com/samdenty/gqless/blob/master/gqless/src/Node/ArrayNode.ts#L77 it has to be changed to:

```ts
return arr?.length ?? 0
```

In my tests it worked but I'm not sure if it's not gonna mess something else.

**Comment by lukejagodzinski** [comment]
@samdenty I've seen your comment in PR that empty `ArrayNode`'s `length` has to be `1` to prevent multiple requests. So what would be solution for that problem? It's also causing another problem when let's say I have a list of elements that are being removed one by one by mutation and after each mutation I refetch query then it will work nicely for more than one element on the list but if I get to the last element, even though server will return an empty list, it won't be updated in the cache (or maybe it is but ArrayNode length is still 1 so it still displays this element that was removed.

**Comment by samdenty** [comment]
We need to render the Story component, so we can see what data it uses. This is why it's set to 1, before the data has been fetched.

Once the data's been fetched, it'll return the actual length - which may be empty.

I'll get back to you later, but updating arrays in the cache hasn't been fully tested yet. It's most likely a cache problem you're running into.

**Comment by lukejagodzinski** [comment]
Yep I understand reasoning behind having length 1. It totally makes sense. While I'm on it I will try to play with cache and just understand how everything work under the hood, so maybe I will be able to come up with some PR. Thanks for the fast response!

**Comment by lukejagodzinski** [comment]
Ok I've checked cache with empty arrays and everything works well there, so no need for changes. However, I found another issue that is not related to the `length = 1` issue. There is a problem in the `refetch` helper. I've confirmed it in my tests. Co here is how to reproduce it (sorry for no reproduction repo):

1. Display list of elements, let's say `query.users`. Let's say that initial query returns 5 elements
2. Manually change in DB number of users so that the `query.users` query will return different number of elements now
3. Run refetch for the `query.users` query
4. Component will rerender but will the old list of elements even though that cache was properly updated
5. Run refetch again and now component will render with appropriate number of elements
